google/cloud/spanner/database_admin_client.cc

```diff
@@ -64,6 +64,57 @@ StatusOr<google::iam::v1::Policy> DatabaseAdminClient::SetIamPolicy(
   return conn_->SetIamPolicy({std::move(db), std::move(policy)});
 }
 
+StatusOr<google::iam::v1::Policy> DatabaseAdminClient::SetIamPolicy(
+    Database const& db, IamUpdater const& updater) {
+  auto const rerun_maximum_duration = std::chrono::minutes(15);
+  auto default_rerun_policy =
+      LimitedTimeTransactionRerunPolicy(rerun_maximum_duration).clone();
+
+  auto const backoff_initial_delay = std::chrono::milliseconds(1000);
+  auto const backoff_maximum_delay = std::chrono::minutes(5);
+  auto const backoff_scaling = 2.0;
+  auto default_backoff_policy =
+      ExponentialBackoffPolicy(backoff_initial_delay, backoff_maximum_delay,
+                               backoff_scaling)
+          .clone();
+
+  return SetIamPolicy(db, updater, std::move(default_rerun_policy),
+                      std::move(default_backoff_policy));
+}
+
+StatusOr<google::iam::v1::Policy> DatabaseAdminClient::SetIamPolicy(
+    Database const& db, IamUpdater const& updater,
+    std::unique_ptr<TransactionRerunPolicy> rerun_policy,
+    std::unique_ptr<BackoffPolicy> backoff_policy) {
+  using RerunnablePolicy = internal::SafeTransactionRerun;
+
+  Status last_status =
+      Status(StatusCode::kUnknown, "Exhausted before first call");
+  for (; !rerun_policy->IsExhausted();
+       std::this_thread::sleep_for(backoff_policy->OnCompletion())) {
+    auto current_policy = GetIamPolicy(db);
+    if (!current_policy) {
+      last_status = std::move(current_policy).status();
+    } else {
+      auto etag = current_policy->etag();
+      auto desired = updater(*current_policy);
+      if (!desired.has_value()) {
+        return current_policy;
+      }
+      desired->set_etag(std::move(etag));
+      auto result = SetIamPolicy(db, *std::move(desired));
+      if (RerunnablePolicy::IsOk(result.status())) {
+        return result;
+      }
+      last_status = std::move(result).status();
+    }
+    if (!rerun_policy->OnFailure(last_status)) {
```

Comment:
Nit: It is a little funny that `!OnFailure(s)` means `IsPermanentFailure(s) || IsExhausted()`, so, if we continue, we know we saw `!IsExhausted()`.  Yet, the very next thing we test is `!IsExhausted()`.  For a "limited count" policy this is redundant.  For a "limited time" policy it means we're immediately sampling the clock again, which feels redundant/racy.

We could address that, but the bigger question might be whether any rerun policy should be able to trigger an "Exhausted before first call".  After all, it isn't really a "rerun" if it was never a "run".

Reply:
I think it is funny too.  WDTY of the do/while loop instead?

Reply:
I like it better.  But things like ...
```
  do {
   ...
    if (!rerun_policy->OnFailure(last_status)) break;
    std::this_thread::sleep_for(backoff_policy->OnCompletion());
  } while (!rerun_policy->IsExhausted());
```
reinforce the idea that the rerun and backoff policies should be unified.  It is no use sleeping beyond the deadline, for example.

The main feature would be that `rerun_policy->OnFailure(status)` would either return `false` _immediately_ on a permanent error, or `false` _when_ the count/deadline has been exceeded, or `true` _when_ the next attempt should be made.

But that is something for another day.